### PR TITLE
LibTimeZone: Allow users to change the system time zone :^)

### PR DESCRIPTION
**Pull request by trflynn89**
tl;dr:
tz

This moves LibTimeZone to be apart of LibC, as LibC directly requires its data for the `time.h` functions like `localtime`. Then updates LibTimeZone to use `/etc/timezone` to retrieve/store the system time zone, and adds command line / GUI applications to do the same.

This only updates `localtime` to use the system time zone as a first step, which is enough for the system clock. There's a few time zone related FIXMEs in time.h already that we should be able to address shortly.

In the future, we may remove LibTimeZone's compiled data in favor of a set of data files added to the system to be read at runtime. But I'm leaving that for another PR.

### CODE REVIEW

**Comment by trflynn89** [comment]
Edit: rebased just to make sure we pass in all time zones still after 9cd93944b8c7363d4654c6239d317aed31fa88e2

File: Userland/Libraries/LibTimeZone/TimeZone.cpp [hunk 1]
```diff
@@ -20,15 +22,77 @@ enum class TimeZone : u16 {
...
 StringView current_time_zone()
 {
-    static bool initialized_time_zone = false;
-    if (!initialized_time_zone) {
-        initialized_time_zone = true;
-        tzset();
-    }
+    TimeZoneFile time_zone_file("r");
+
+    // FIXME: Propogate the error to existing callers.
```
Comment (alimpfard) [inline Userland/Libraries/LibTimeZone/TimeZone.cpp:76]: Typo: Propogate => Propagate

File: Userland/Libraries/LibTimeZone/TimeZone.cpp [hunk 2]
```diff
@@ -20,15 +22,77 @@ enum class TimeZone : u16 {
...
+{
+#ifdef __serenity__
+    TimeZoneFile time_zone_file("w");
+
+    if (auto new_time_zone = canonicalize_time_zone(time_zone); new_time_zone.has_value())
+        return time_zone_file.write_time_zone(*new_time_zone);
+
+    return Error::from_string_literal("Provided time zone is not supported");
+#else
+    // Do not event attempt to change the time zone of someone's host machine.
```
Comment (alimpfard) [inline Userland/Libraries/LibTimeZone/TimeZone.cpp:93]: Typo: event => even
Suggestion: someone => the user

File: Userland/Utilities/fortune.cpp [hunk 3]
```diff
@@ -84,7 +84,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     auto file = TRY(Core::File::open(path, Core::OpenMode::ReadOnly));
 
-    TRY(Core::System::pledge("stdio"));
```
Comment (alimpfard) [inline Userland/Utilities/fortune.cpp:87]: Why is the pledge gone?

File: Userland/Utilities/timezone.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    TRY(Core::System::pledge("stdio rpath wpath cpath"));
+    TRY(Core::System::unveil("/etc/timezone", "rwc"));
+    TRY(Core::System::unveil(nullptr, nullptr));
+
+    StringView time_zone;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(time_zone, "The time zone to set", "file", Core::ArgsParser::Required::No);
```
Comment (alimpfard) [inline Userland/Utilities/timezone.cpp:23]: The timezone is not exactly a file (in the user's perspective anyway)

**Review by alimpfard — COMMENTED** [review]
n i t p i c k s

File: Userland/Utilities/fortune.cpp (hunk 3 above)
Comment (trflynn89) [reply Userland/Utilities/fortune.cpp:87]: Diff is kinda hard to tell, but we have a pledge line above this:
`TRY(Core::System::pledge("stdio rpath"));`
This commit was to add rpath, so deleting this line is just keeping rpath in fortune.

File: Userland/Utilities/timezone.cpp (hunk 4 above)
Comment (trflynn89) [reply Userland/Utilities/timezone.cpp:23]: Oops, copy-paste mistake

**Review by linusg — COMMENTED** [review]
Very nice :^) Couple of questions though:

> We will use /etc/Clock.ini for system-wide clock settings, which as of this commit, consists of just the time zone.

Huh? I don't see any references to this file... Don't we use `/etc/timezone`?

> Userland: Add promises to programs that will read /etc/timezone

I feel like this commit should be squashed into the previous one, otherwise the affected applications may stop working when bisecting.

File: Userland/Libraries/LibTimeZone/TimeZone.cpp [hunk 5]
```diff
@@ -20,15 +22,77 @@ enum class TimeZone : u16 {
 };
 #endif
 
+class TimeZoneFile {
+public:
+    TimeZoneFile(char const* mode)
+        : m_file(fopen("/etc/timezone", mode))
```
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibTimeZone/TimeZone.cpp:28]: > Huh? I don't see any references to this file... Don't we use `/etc/timezone`?

Oops! That was a dangling reference to an old impl of this PR, good catch.
 
> I feel like this commit should be squashed into the previous one, otherwise the affected applications may stop working when bisecting.

I'll move that commit to be one earlier - my initial thought was that the commit order didn't matter there, because the LibC method wasn't using /etc/timezone yet, but on closer look that isn't quite right, WebContent / Browser would get tripped up.

**Comment by trflynn89** [comment]
> Huh? I don't see any references to this file... Don't we use `/etc/timezone`?

Oops! That was a dangling reference to an old impl of this PR, good catch.
 
> I feel like this commit should be squashed into the previous one, otherwise the affected applications may stop working when bisecting.

I'll move that commit to be one earlier - my initial thought was that the commit order didn't matter there, because the LibC method wasn't using /etc/timezone yet, but on closer look that isn't quite right, WebContent / Browser would get tripped up.

File: Userland/Libraries/LibTimeZone/TimeZone.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibTimeZone/TimeZone.cpp:28]: Right, so tbh I'm not 100% what the right thing to do here is yet. I'm not expecting host systems to have /etc/timezone. But my (limited) understanding of this stuff is that if your program cares about time zones, then it should `tzset` at startup. That method should check the `TZ` var, then fall back to a system-dependent default. My intention is that reading this file is the system-dependent default for Serenity. Then, eventually, the `localtime` impl here will respect what `tzset` set up.

I've held off on implementing that yet because I'm not quite sure how that allows for reacting to time zone changes. This current impl allows for changes, but I'm sure needs to be thought about further.
